PandasConnector

rhunwicks opened the pull request:
Here is an initial draft of PandasConnector for consideration, as discussed in #3302.

It is not complete, but it shows the general approach.

I have created contrib.tests.connectors which implements a BaseConnectorTestCase which I am using to make sure that the results returned by PandasConnector are consistent with those from SqlaTable.

If we get to the point where this PR can be merged, then we will probably want to refactor to put BaseConnectorTestCase and SqlaConnectorTestCase into the main tests directory and just leave the PandasConnectorTestCase in the contrib directory.

rhunwicks committed:
Initial commit of PandasConnector - see #3302

rhunwicks committed:
remove temporary files - see #3302

rhunwicks committed:
Fix PandasDatasource.baselink - see #3302

xrmx commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,724 @@
+
+class PandasDatabase(object):
+    """Non-ORM object for a Pandas Source"""
+    database_name = ''
```
i don't think you need to set defaults if they are not optional in __init__

xrmx commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,724 @@
+
+    Each Pandas Datasource can have multiple columns"""
+
+    __tablename__ = 'pandascolumns'
```
We may want to be consistent with sqla backend and call it  'pandas_columns'

xrmx commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,724 @@
+        # Build a dict of the metrics to include, including those that
+        # are required for post-aggregation filtering
+        filtered_metrics = [flt['col']
+                            for flt in extras.get('having_druid', [])
```
having_druid? If we are reusing the field a comment would be nice

rhunwicks replied:
I'm adding a comment for the use of `having_druid`, and similar ones for `granularity_sqla` and `time_grain_sqla`. However, if there is a long term plan to encourage the development of more connectors it would be better to rename these parameters in the main code. For example to `granularity_col`, `granularity_freq` and `post_aggregation_filter`.

xrmx commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,724 @@
+
+        This is used to be displayed to the user so that she/he can
+        understand what is taking place behind the scene"""
+        import json
```
you can make these imports global

rhunwicks replied:
I'll take them out, they are part of the debugging code.

xrmx commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,724 @@
+            return val.isoformat() + "Z"
+
+        logging.info(json.dumps(query_obj, indent=4, default=to_serializable))
+        print(json.dumps(query_obj, indent=4, default=to_serializable))
```
debug

xrmx reviewed (commented):
Some small nitpicks

xrmx commented on contrib/connectors/pandas/views.py:
```diff
@@ -0,0 +1,270 @@
+        'format': _(
+            "The format of the raw data, e.g. csv"),
+        'additional_parameters': _(
+            "A JSON-formatted dictionary of additional parameters "
```
Could we have add something like: "These are the actual parameters for the read_* functions, see https://pandas.pydata.org/pandas-docs/stable/api.html"

rhunwicks replied:
Done

xrmx commented:
Thanks a lot for your great work, i'd really like this to be a first citizen in superset

rhunwicks committed:
Rename PandasConnectors sqla tables for consistency

rhunwicks committed:
Remove redundant debug code

rhunwicks commented on contrib/connectors/pandas/models.py:
```diff
@@ -0,0 +1,705 @@
+        # for post-aggregation filters, and we are reusing that
+        # interface component.
+        filtered_metrics = [flt['col']
+                            for flt in extras.get('having_druid', [])
```
I'm have added a comment for the use of `having_druid`, and similar ones for `granularity_sqla` and `time_grain_sqla`. However, if there is a long term plan to encourage the development of more connectors it would be better to rename these parameters in the main code. For example to `granularity_col`, `granularity_freq` and `post_aggregation_filter`.

rhunwicks committed:
Update PandasConnector for order_desc - see #3302

rhunwicks committed:
Add PandasConnector migration to main versions folder - see #3302

rhunwicks committed:
Add PandasConnector migration to main versions folder - see #3302

rhunwicks committed:
Merge remote-tracking branch 'upstream/master' into 3302-pandas-connctor

rhunwicks committed:
Add PandasConnector migration to main versions folder - see #3302

rhunwicks committed:
Renamed connector_tests.py - see #3302

rhunwicks committed:
Add lxml and bs4 to dev-reqs - see #3302

rhunwicks committed:
Update PandasMetric with warning_text - see #3302

rhunwicks committed:
More efficient PandasDatasource.get_metadata - see #3302

rhunwicks committed:
Python27 compatibility for PandasConnector - see #3302

rhunwicks commented:
@mistercrunch I have moved the migration into `superset/migrations/versions` in order to get Travis to run. I can move it back out if you decide you want to keep it in a `contrib` directory.

I have also added `lxml` and `beautifulsoup4` to `dev-reqs.txt` because `contrib.tests.connector_tests` uses them to prepare test data. `connector_tests` is probably useful for other people writing connectors as well as providing increased test coverage for `superset/connectors/sqla/models.py` so you might want to adopt some of it into core.

Please let me know what you would like me to do next?

xrmx commented on dev-reqs.txt:
```diff
@@ -1,7 +1,10 @@
 coveralls
 flake8
 flask_cors
+lxml==3.8.0
```
once should be enough

xrmx replied:
actually i can't find where you are using lxml and beautifulsoup

rhunwicks replied:
The are required by `pandas.read_html()`, which I use in `connector_tests` to make it easy to set up test data as a Markdown table and then push it into whatever Datasource is being tested.

xrmx replied:
Then they should be added to setup.py or we are going to miss them in production :)

rhunwicks replied:
https://pandas.pydata.org/pandas-docs/stable/io.html#io-html-gotchas explains that `read_html` uses bs4 as a wrapper for lxml. We could use `html5lib` instead of `lxml`.

xrmx replied:
The original comment was because you added lxml twice not because you added two different libraries :)

rhunwicks replied:
> The original comment was because you added lxml twice

Thanks - I've fixed it now.

rhunwicks replied:
> Then they should be added to setup.py or we are going to miss them in production

We only *need* them for `connector_tests`. I am assuming that we don't need to be able to run the tests in production (hence `dev-reqs.txt`). You would only need them in production if you wanted to be able to use an html-format PandasDatasource. But there are many possible extra libraries that we might need to support the full range of `read_xxx` Pandas functionality - hdf5, xlrd, probably something for Stata. I was assuming that we wouldn't put all the possible dependencies into Superset, and we would expect the individual implementations to install the correct libraries for the formats they want/need to support.

If it is preferable that Superset supports the full range of Pandas `read_xxx` out of the box then I could update `setup.py` to include them and add tests. But that only makes sense if PandasConnector is going to be in core rather than in a `contrib` directory.

xrmx replied:
If html is a format supported, wouldn't parse_html used then?

rhunwicks replied:
@xrmx - you are correct - we are specifically supporting html in `FORMATS` so I need to add those dependencies to `setup.py` or whatever requirements file we ship with PandasConnector if it lives outside core.

I wonder if there are any other formats we should support by default? JSON should work by default. Excel might be useful for many people and would need `xlrd`? Given the normal Superset audience do we want to support HDF5 and/or Feather? What about SAS or Stata?

xrmx replied:
In my opinion we should keep the pandas data source only for the simple cases: excel, csv, html and json. For more complex scenarios i think an apache arrow backend as suggested by @mistercrunch is the way to go.

rhunwicks replied:
I agree that Arrow is the future. http://wesmckinney.com/blog/apache-arrow-pandas-internals/ is a good read.

rhunwicks replied:
I have added the necessary packages to `setup.py` so that `PandasDatasource` works for csv, html, json and Excel.

@xrmx do you think it would be a good idea to allow `PandasDatasource.FORMATS` to be overridden from settings, much as we allow `ADDITIONAL_MODULE_DS_MAP` to add extra Connectors? Then a specific installation could, for example, install `pytables` and then set:

```python
ADDITIONAL_PANDAS_FORMATS = [('hdf', 'HDF5')]
```

Another approach would be to attempt to dynamically set the list of allowable formats based on whether we can import the required dependencies.

xrmx replied:
Uhm, we probably don't want to tie the supported formats to the db so maybe it's better to just have a varchar in the db and limit the values from a select built from a config key so each user can support whatever formats they want.

rhunwicks committed:
Remove duplicate lxml from dev-reqs.txt - see #3302

rhunwicks committed:
Use ChoiceType for PandasDatasource.format - see #3302

rhunwicks committed:
Use ChoiceType for PandasDatasource.format - see #3302

rhunwicks committed:
Allow calculated PandasColumns and multi-column PandasMetrics - see #3302

rhunwicks committed:
Support additional Pandas formats if dependencies are satisfied - see #3302

rhunwicks committed:
Flake8 fixes - see #3302

rhunwicks committed:
Better tests for summary metrics - see #3302

rhunwicks committed:
Cache source dataframes in PandasDatasource - see #3302

rhunwicks committed:
Py2 fix for DataFrameCache - see #3302

rhunwicks committed:
Py2 fix for DataFrameCache - see #3302

rhunwicks committed:
Py2 fix for DataFrameCache - see #3302

rhunwicks committed:
Configure dataframe cache from settings - see #3302

rhunwicks committed:
Fix import in contrib.tests.cache_tests - see #3302

rhunwicks committed:
Merge branch 'master' into 3302-pandas-connctor

rhunwicks committed:
Merge branch 'master' into 3302-pandas-connctor

rhunwicks committed:
Support authentication against remote APIs for PandasConnector - see
#3302

rhunwicks committed:
Flake8 fixes - see #3302

rhunwicks committed:
Flake8 fixes - see #3302

rhunwicks committed:
Improved dataframe cache - see #3302

rhunwicks committed:
Restore datasource count metric for PandasDatasource - see #3302

rhunwicks committed:
Include dttm_cols in datasource.data - see #3302

rhunwicks committed:
Put deps at end of setup.py to reduce merge conflicts - see #3302

rhunwicks committed:
Merge remote-tracking branch 'upstream/master' into 3302-pandas-connctor

rhunwicks committed:
Flake8 fix for setup.py - see #3302

rhunwicks commented:
Having the migrations and requirements mixed in with the main Superset ones without ever merging this MR causes frequent merge conflicts. Therefore, I am going to close this MR. We will continue to maintain the connector in order to allow us to use Superset with APIs and remote files, but we will do so in our fork rather than a MR.

stu-co commented:
@rhunwicks thank you for creating this code. It seems pretty perfect for us to be honest -  I wish it had made it into superset!
